packages/ts-transformers/src/opaque-ref/dependency.ts

```diff
@@ -0,0 +1,730 @@
+      };
+    }
+
+    if (ts.isArrowFunction(expression) || ts.isFunctionExpression(expression)) {
+      const parameterSymbols: ts.Symbol[] = [];
+      for (const parameter of expression.parameters) {
+        const symbol = checker.getSymbolAtLocation(parameter.name);
+        if (symbol) parameterSymbols.push(symbol);
+      }
+      const childScope = createScope(context, scope, parameterSymbols);
+      if (ts.isBlock(expression.body)) {
+        const analyses: InternalAnalysis[] = [];
+        for (const statement of expression.body.statements) {
+          if (ts.isReturnStatement(statement) && statement.expression) {
+            analyses.push(
+              analyzeExpression(statement.expression, childScope, context),
+            );
+          }
+        }
+        return mergeAnalyses(...analyses);
+      }
+      return analyzeExpression(expression.body, childScope, context);
+    }
+
+    if (ts.isObjectLiteralExpression(expression)) {
+      const analyses = expression.properties.map((prop) => {
+        if (
+          ts.isPropertyAssignment(prop) && ts.isExpression(prop.initializer)
+        ) {
+          return analyzeExpression(prop.initializer, scope, context);
+        }
+        if (ts.isShorthandPropertyAssignment(prop)) {
+          return analyzeExpression(prop.name, scope, context);
+        }
+        return emptyAnalysis();
+      });
+      return mergeAnalyses(...analyses);
+    }
+
+    if (ts.isArrayLiteralExpression(expression)) {
+      const analyses = expression.elements.map((element) => {
+        if (ts.isExpression(element)) {
+          return analyzeExpression(element, scope, context);
+        }
+        return emptyAnalysis();
+      });
+      return mergeAnalyses(...analyses);
+    }
+
+    const analyses: InternalAnalysis[] = [];
+    expression.forEachChild((child) => {
+      if (ts.isExpression(child)) {
+        analyses.push(analyzeExpression(child, scope, context));
+      }
+    });
+    if (analyses.length === 0) {
+      return emptyAnalysis();
+    }
+    return mergeAnalyses(...analyses);
+  };
+
+  return (expression: ts.Expression) => {
+    const context: AnalyzerContext = {
+      nextNodeId: 0,
+      nextScopeId: 0,
+      nodes: [],
+      scopes: new Map(),
+    };
+    const rootScope = createScope(context, null, []);
+    const result = analyzeExpression(expression, rootScope, context);
+    const scopes = Array.from(context.scopes.values()).map(toDependencyScope);
+    return {
+      ...result,
+      graph: {
+        nodes: context.nodes,
+        scopes,
+        rootScopeId: rootScope.id,
+      },
+    };
+  };
+}
+
+export function collectOpaqueRefs(
+  node: ts.Node,
+  checker: ts.TypeChecker,
+): ts.Expression[] {
+  const refs: ts.Expression[] = [];
+  const processedNodes = new Set<ts.Node>();
+
+  const visit = (n: ts.Node): void => {
+    if (ts.isJsxAttribute(n)) {
+      const name = n.name.getText();
+      if (name && name.startsWith("on")) {
```

Comment:
it's not clear to me why the event handlers (attributes that start with "on") are special-cased here, maybe worth a comment